syntax = "proto3";
package librevault.serialization;

message AES_CBC {
	bytes ct = 1;
	bytes iv = 2;
}

message Meta {
	/* Required metadata */
	bytes path_id = 1;
	AES_CBC path = 2;	// AES-256-CBC encrypted normalized path relative to sync directory path. It cannot contain dot (".") and dot-dot ("..") elements. Directory path MUST NOT end with trailing slash ("/"). Separators MUST be slashes ("/").
	uint32 meta_type = 3;
	int64 revision = 4;

	/* Content-specific metadata */
	message GenericMetadata {	// Filesystem attributes (ALL AND NOT DELETED)
    	int64 mtime = 1;
    	// Windows-specific
    	uint32 windows_attrib = 2;
    	// Unix-specific
    	uint32 mode = 3;
        uint32 uid = 4;
        uint32 gid = 5;
	}
	message SymlinkMetadata {
    	AES_CBC symlink_path = 1;
	}

	message FileMetadata {
		// Algorithm selection
		uint32 algorithm_type = 1;
		uint32 strong_hash_type = 2;

		// Uni-algorithm parameters
		uint32 max_chunksize = 3;
		uint32 min_chunksize = 4;

		// Rabin algorithm parameters
		message RabinGlobalParams {
			uint64 polynomial = 1;
			uint32 polynomial_degree = 2;
			uint32 polynomial_shift = 3;
			uint32 avg_bits = 4;
		}
		AES_CBC rabin_global_params = 5;

		message Chunk {
			bytes ct_hash = 1;
			uint32 size = 2;
			bytes iv = 3;

			bytes pt_hmac = 4;
		}
		repeated Chunk chunks = 255;
	}

	GenericMetadata generic_metadata = 16;
	oneof type_specific_metadata {
		SymlinkMetadata symlink_metadata = 17;
		FileMetadata file_metadata = 18;
	}
}
